File: csharp/src/Core.Workflows/Activities/UtilityActivities.cs
﻿using Temporalio.Activities;
using Train.Solver.Core.Abstractions.Models;

namespace Train.Solver.Core.Workflows.Activities;


public class UtilityActivities
{
    [Activity]
    public IEnumerable<BlockRangeModel> GenerateBlockRanges(ulong start, ulong end, uint chunkSize)
    {
        if (chunkSize == 0)
            throw new ArgumentException("Max size must be greater than 0", nameof(chunkSize));

        var result = new List<BlockRangeModel>();

        ulong currentStart = start;

        while (currentStart <= end)
        {
            ulong currentEnd = Math.Min(currentStart + chunkSize - 1, end);
            result.Add(new BlockRangeModel(currentStart, currentEnd));
            currentStart = currentEnd + 1;
        }

        return result;
    }
}

File: js/src/Blockchian/Blockchain.Abstraction/Activities/UtilityActivities.ts
import { BlockRangeModel } from "../Models/BlockRangeModel";

export function GenerateBlockRanges(start: number, end: number, chunkSize: number): BlockRangeModel[] {
    if (chunkSize <= 0) {
        throw new Error('Chunk size must be greater than 0');
    }

    const result: BlockRangeModel[] = [];
    let currentStart = start;

    while (currentStart <= end) {
        const currentEnd = Math.min(currentStart + chunkSize - 1, end);
        result.push({ From: currentStart, To: currentEnd });
        currentStart = currentEnd + 1;
    }

    return result;
}
